rpc: getnetmsgstats phase II

satsie opened the pull request:
**This PR is now outdated! Please see #4 for the latest**

----------------------------------------------------

Phase II of my work for https://github.com/bitcoin/bitcoin/issues/26337. 

This  reflects the implementation for a new `getnetmsgstats` RPC, which supports several kinds of filters (msgtype, conntype, network, etc.) in any order.

When I am done working through everything I plan to redo this entire PR for the feature, specifically to rework the commit history.

To test:

1. Build (`make -j 4` for me.. I know 4 is a weird number but my computer is kind of temperamental)
2. `./src/bitcoind -debug=net`
3. In another terminal run: `./src/bitcoin-cli getnetmsgstats`

Examples:

<details>
  <summary>getnetmsgstats</summary>

```
 ./src/bitcoin-cli getnetmsgstats
{
  "sent": {
    "total": {
      "msg_count": 175,
      "total_bytes": 34109
    }
  },
  "received": {
    "total": {
      "msg_count": 280,
      "total_bytes": 2144499
    }
  }
}
```
</details>

<details>
<summary>getnetmsgstats '["conntype","msgtype"]'</summary>

```
./src/bitcoin-cli getnetmsgstats '["conntype","msgtype"]'
{
  "sent": {
    "block-relay-only": {
      "getheaders": {
        "msg_count": 5,
        "total_bytes": 5265
      },
      "headers": {
        "msg_count": 4,
        "total_bytes": 424
      },
      "inv": {
        "msg_count": 1,
        "total_bytes": 61
      },
      "ping": {
        "msg_count": 5,
        "total_bytes": 160
      },
      "pong": {
        "msg_count": 3,
        "total_bytes": 96
      },
      "sendaddrv2": {
        "msg_count": 4,
        "total_bytes": 96
      },
      "sendcmpct": {
        "msg_count": 5,
        "total_bytes": 165
      },
      "sendheaders": {
        "msg_count": 2,
        "total_bytes": 48
      },
      "verack": {
        "msg_count": 5,
        "total_bytes": 120
      },
      "version": {
        "msg_count": 5,
        "total_bytes": 635
      },
      "wtxidrelay": {
        "msg_count": 4,
        "total_bytes": 96
      }
    },
    "outbound-full-relay": {
      "addr": {
        "msg_count": 3,
        "total_bytes": 165
      },
      "addrv2": {
        "msg_count": 10,
        "total_bytes": 452
      },
      "feefilter": {
        "msg_count": 9,
        "total_bytes": 288
      },
      "getaddr": {
        "msg_count": 9,
        "total_bytes": 216
      },
      "getblocktxn": {
        "msg_count": 2,
        "total_bytes": 4027
      },
      "getdata": {
        "msg_count": 120,
        "total_bytes": 24312
      },
      "getheaders": {
        "msg_count": 9,
        "total_bytes": 9477
      },
      "headers": {
        "msg_count": 11,
        "total_bytes": 1004
      },
      "inv": {
        "msg_count": 222,
        "total_bytes": 87378
      },
      "ping": {
        "msg_count": 9,
        "total_bytes": 288
      },
      "pong": {
        "msg_count": 9,
        "total_bytes": 288
      },
      "sendaddrv2": {
        "msg_count": 7,
        "total_bytes": 168
      },
      "sendcmpct": {
        "msg_count": 11,
        "total_bytes": 363
      },
      "sendheaders": {
        "msg_count": 8,
        "total_bytes": 192
      },
      "tx": {
        "msg_count": 14,
        "total_bytes": 3510
      },
      "verack": {
        "msg_count": 9,
        "total_bytes": 216
      },
      "version": {
        "msg_count": 10,
        "total_bytes": 1270
      },
      "wtxidrelay": {
        "msg_count": 7,
        "total_bytes": 168
      }
    }
  },
  "received": {
    "block-relay-only": {
      "feefilter": {
        "msg_count": 2,
        "total_bytes": 64
      },
      "getheaders": {
        "msg_count": 3,
        "total_bytes": 3159
      },
      "headers": {
        "msg_count": 3,
        "total_bytes": 237
      },
      "inv": {
        "msg_count": 1,
        "total_bytes": 133
      },
      "ping": {
        "msg_count": 3,
        "total_bytes": 96
      },
      "pong": {
        "msg_count": 3,
        "total_bytes": 96
      },
      "sendaddrv2": {
        "msg_count": 3,
        "total_bytes": 72
      },
      "sendcmpct": {
        "msg_count": 5,
        "total_bytes": 165
      },
      "sendheaders": {
        "msg_count": 3,
        "total_bytes": 72
      },
      "verack": {
        "msg_count": 5,
        "total_bytes": 120
      },
      "version": {
        "msg_count": 5,
        "total_bytes": 631
      },
      "wtxidrelay": {
        "msg_count": 3,
        "total_bytes": 72
      }
    },
    "outbound-full-relay": {
      "addr": {
        "msg_count": 4,
        "total_bytes": 60164
      },
      "addrv2": {
        "msg_count": 15,
        "total_bytes": 110294
      },
      "blocktxn": {
        "msg_count": 2,
        "total_bytes": 4600726
      },
      "cmpctblock": {
        "msg_count": 3,
        "total_bytes": 37593
      },
      "feefilter": {
        "msg_count": 9,
        "total_bytes": 288
      },
      "getdata": {
        "msg_count": 11,
        "total_bytes": 779
      },
      "getheaders": {
        "msg_count": 9,
        "total_bytes": 9477
      },
      "headers": {
        "msg_count": 14,
        "total_bytes": 1484
      },
      "inv": {
        "msg_count": 135,
        "total_bytes": 61155
      },
      "notfound": {
        "msg_count": 1,
        "total_bytes": 277
      },
      "ping": {
        "msg_count": 9,
        "total_bytes": 288
      },
      "pong": {
        "msg_count": 9,
        "total_bytes": 288
      },
      "sendaddrv2": {
        "msg_count": 8,
        "total_bytes": 192
      },
      "sendcmpct": {
        "msg_count": 16,
        "total_bytes": 528
      },
      "sendheaders": {
        "msg_count": 9,
        "total_bytes": 216
      },
      "tx": {
        "msg_count": 583,
        "total_bytes": 395235
      },
      "verack": {
        "msg_count": 10,
        "total_bytes": 240
      },
      "version": {
        "msg_count": 10,
        "total_bytes": 1276
      },
      "wtxidrelay": {
        "msg_count": 8,
        "total_bytes": 192
      }
    }
  }
}

```
</details>

<details>
<summary>getnetmsgstats '["conntype", "network", "msgtype"]'</summary>

```
./src/bitcoin-cli getnetmsgstats '["conntype", "network", "msgtype"]'
{
  "sent": {
    "block-relay-only": {
      "ipv4": {
        "getheaders": {
          "msg_count": 5,
          "total_bytes": 5265
        },
        "headers": {
          "msg_count": 4,
          "total_bytes": 424
        },
        "inv": {
          "msg_count": 1,
          "total_bytes": 61
        },
        "ping": {
          "msg_count": 6,
          "total_bytes": 192
        },
        "pong": {
          "msg_count": 4,
          "total_bytes": 128
        },
        "sendaddrv2": {
          "msg_count": 4,
          "total_bytes": 96
        },
        "sendcmpct": {
          "msg_count": 5,
          "total_bytes": 165
        },
        "sendheaders": {
          "msg_count": 2,
          "total_bytes": 48
        },
        "verack": {
          "msg_count": 5,
          "total_bytes": 120
        },
        "version": {
          "msg_count": 5,
          "total_bytes": 635
        },
        "wtxidrelay": {
          "msg_count": 4,
          "total_bytes": 96
        }
      }
    },
    "outbound-full-relay": {
      "ipv4": {
        "addr": {
          "msg_count": 6,
          "total_bytes": 360
        },
        "addrv2": {
          "msg_count": 15,
          "total_bytes": 868
        },
        "feefilter": {
          "msg_count": 9,
          "total_bytes": 288
        },
        "getaddr": {
          "msg_count": 9,
          "total_bytes": 216
        },
        "getblocktxn": {
          "msg_count": 2,
          "total_bytes": 4027
        },
        "getdata": {
          "msg_count": 210,
          "total_bytes": 42978
        },
        "getheaders": {
          "msg_count": 9,
          "total_bytes": 9477
        },
        "headers": {
          "msg_count": 11,
          "total_bytes": 1004
        },
        "inv": {
          "msg_count": 431,
          "total_bytes": 175079
        },
        "ping": {
          "msg_count": 17,
          "total_bytes": 544
        },
        "pong": {
          "msg_count": 17,
          "total_bytes": 544
        },
        "sendaddrv2": {
          "msg_count": 7,
          "total_bytes": 168
        },
        "sendcmpct": {
          "msg_count": 11,
          "total_bytes": 363
        },
        "sendheaders": {
          "msg_count": 8,
          "total_bytes": 192
        },
        "tx": {
          "msg_count": 59,
          "total_bytes": 41631
        },
        "verack": {
          "msg_count": 9,
          "total_bytes": 216
        },
        "version": {
          "msg_count": 10,
          "total_bytes": 1270
        },
        "wtxidrelay": {
          "msg_count": 7,
          "total_bytes": 168
        }
      }
    }
  },
  "received": {
    "block-relay-only": {
      "ipv4": {
        "feefilter": {
          "msg_count": 2,
          "total_bytes": 64
        },
        "getheaders": {
          "msg_count": 3,
          "total_bytes": 3159
        },
        "headers": {
          "msg_count": 3,
          "total_bytes": 237
        },
        "inv": {
          "msg_count": 1,
          "total_bytes": 133
        },
        "ping": {
          "msg_count": 4,
          "total_bytes": 128
        },
        "pong": {
          "msg_count": 4,
          "total_bytes": 128
        },
        "sendaddrv2": {
          "msg_count": 3,
          "total_bytes": 72
        },
        "sendcmpct": {
          "msg_count": 5,
          "total_bytes": 165
        },
        "sendheaders": {
          "msg_count": 3,
          "total_bytes": 72
        },
        "verack": {
          "msg_count": 5,
          "total_bytes": 120
        },
        "version": {
          "msg_count": 5,
          "total_bytes": 631
        },
        "wtxidrelay": {
          "msg_count": 3,
          "total_bytes": 72
        }
      }
    },
    "outbound-full-relay": {
      "ipv4": {
        "addr": {
          "msg_count": 5,
          "total_bytes": 60219
        },
        "addrv2": {
          "msg_count": 23,
          "total_bytes": 110809
        },
        "blocktxn": {
          "msg_count": 2,
          "total_bytes": 4600726
        },
        "cmpctblock": {
          "msg_count": 3,
          "total_bytes": 37593
        },
        "feefilter": {
          "msg_count": 9,
          "total_bytes": 288
        },
        "getdata": {
          "msg_count": 42,
          "total_bytes": 3174
        },
        "getheaders": {
          "msg_count": 9,
          "total_bytes": 9477
        },
        "headers": {
          "msg_count": 14,
          "total_bytes": 1484
        },
        "inv": {
          "msg_count": 243,
          "total_bytes": 109647
        },
        "notfound": {
          "msg_count": 1,
          "total_bytes": 277
        },
        "ping": {
          "msg_count": 17,
          "total_bytes": 544
        },
        "pong": {
          "msg_count": 17,
          "total_bytes": 544
        },
        "sendaddrv2": {
          "msg_count": 8,
          "total_bytes": 192
        },
        "sendcmpct": {
          "msg_count": 16,
          "total_bytes": 528
        },
        "sendheaders": {
          "msg_count": 9,
          "total_bytes": 216
        },
        "tx": {
          "msg_count": 1039,
          "total_bytes": 1225891
        },
        "verack": {
          "msg_count": 10,
          "total_bytes": 240
        },
        "version": {
          "msg_count": 10,
          "total_bytes": 1276
        },
        "wtxidrelay": {
          "msg_count": 8,
          "total_bytes": 192
        }
      }
    }
  }
}
```
</details>



<!--
*** Please remove the following help text before submitting: ***

Pull requests without a rationale and clear improvement may be closed
immediately.

GUI-related pull requests should be opened against
https://github.com/bitcoin-core/gui
first. See CONTRIBUTING.md
-->

<!--
Please provide clear motivation for your patch and explain how it improves
Bitcoin Core user experience or Bitcoin Core developer experience
significantly:

* Any test improvements or new tests that improve coverage are always welcome.
* All other changes should have accompanying unit tests (see `src/test/`) or
  functional tests (see `test/`). Contributors should note which tests cover
  modified code. If no tests exist for a region of modified code, new tests
  should accompany the change.
* Bug fixes are most welcome when they come with steps to reproduce or an
  explanation of the potential issue as well as reasoning for the way the bug
  was fixed.
* Features are welcome, but might be rejected due to design or scope issues.
  If a feature is based on a lot of dependencies, contributors should first
  consider building the system outside of Bitcoin Core, if possible.
* Refactoring changes are only accepted if they are required for a feature or
  bug fix or otherwise improve developer experience significantly. For example,
  most "code style" refactoring changes require a thorough explanation why they
  are useful, what downsides they have and why they *significantly* improve
  developer experience or avoid serious programming bugs. Note that code style
  is often a subjective matter. Unless they are explicitly mentioned to be
  preferred in the [developer notes](/doc/developer-notes.md), stylistic code
  changes are usually rejected.
-->

<!--
Bitcoin Core has a thorough review process and even the most trivial change
needs to pass a lot of eyes and requires non-zero or even substantial time
effort to review. There is a huge lack of active reviewers on the project, so
patches often sit for a long time.
-->

satsie committed:
Create member variables for maps to hold the total bytes received and sent. Initialize message types to zero, and add getters

satsie committed:
Add setters for the new member variable maps

satsie committed:
When messages are received, update the map of received bytes by message type

satsie committed:
When messages are sent, update the map of sent bytes by message type

satsie committed:
Update getnettotals RPC and corresponding functional test

satsie committed:
Run clang-format-diff.py

satsie committed:
Simplify and clean a bunch of variable names

satsie committed:
Pass connection and network types through

satsie commented on src/net.cpp:
```diff
@@ -2642,16 +2651,23 @@ void CConnman::RecordBytesRecv(uint64_t bytes)
 }
 
-void CConnman::RecordBytesRecvByMsgType(mapMsgTypeSize msgtype_bytes)
+void CConnman::RecordMsgStatsRecv(std::map<std::string, std::tuple<int, uint64_t>> msgtype_countbytes,
```
This is where the interesting stuff (so far) comes together

satsie commented on src/net.cpp:
```diff
@@ -2673,8 +2689,12 @@ void CConnman::RecordBytesSent(uint64_t bytes)
 }
 
-void CConnman::RecordBytesSentByMsgType(std::string msg_type, size_t bytes)
+void CConnman::RecordMsgStatsSent(std::string msg_type, size_t bytes, ConnectionType conntype, Network networktype)
```
This is also where the interesting stuff (so far) comes together

satsie committed:
Create and populate the data structures for sent and received network message stats

satsie commented on src/net.h:
```diff
@@ -1002,6 +1004,44 @@ class CConnman
+    struct NetMsgStatsKey {
+        // protocol.cpp has a list of allNetMessageTypes which use the NetMsgType namespace.
+        // Can I treat NetMsgType like an enum?
+        std::string msg_type;
```
Revisit this

amitiuttarwar replied:
it seems like treating the message types works fine & follows the pattern established by the `CNode` message stats. for example , allows reusing the `mapMsgTypeSize` alias. 

unless I'm forgetting something, I don't think it's worth putting in more work to make it an enum

satsie replied:
Good point about being able to reuse things like the `mapMsgTypeSize` alias! Is that what you were referring to when you mentioned `CNodeStats`? I didn't see any `std::string` message type fields on that class. Just want to make sure I'm not missing something. 

`CNodeStats` has two members of type `mapMsgTypeSize`. I assume this is what you were referring to as an example of existing patterns that use message type strings?

Regarding the second half of your comment, at some point you threw out the idea of turning the message types into an enum, but as a separate PR that is completely decoupled from this. I can't recall if the comment I'm replying to here was before or after that discussion. Can you remind me what the latest decision was?

satsie replied:
This comment is no longer relevant now that the workaround for the message type enum is in place (yay!). Closing this comment.

satsie commented on src/net.h:
```diff
@@ -1002,6 +1004,44 @@ class CConnman
+        Network net_type;
+
+        // Is this correct? Is there a better way to write this?
+        bool operator< (const NetMsgStatsKey& rhs) const{
```
`grep -r -i --color=always "operator<("`

Take a look at some other implementations of this overloaded operator

satsie commented on src/net.h:
```diff
@@ -1002,6 +1004,44 @@ class CConnman
+            return false;
+        }
+
+        bool operator= (const NetMsgStatsKey& other) const {
```
Overload `==`? Or just get rid of this? Try changing this to `==` and removing the less than overload

satsie replied:
Removing the overloaded `<` operator resulted in a bunch of compiler errors, even when I corrected this method to overload `==` (instead of `=`). I'm going to remove the `==` overload method.

amitiuttarwar replied:
ah yes, it looks like I misled you. from the docs [here](https://en.cppreference.com/w/cpp/container/map/find) the example points out that the `operator<` is indeed what you need to overload in order to be able to use the `.find` function on a `std::map`.

also, from [here](https://cplusplus.com/reference/map/map/find/)-

> Two keys are considered equivalent if the container's [comparison object](https://cplusplus.com/map::key_comp) returns false reflexively (i.e., no matter the order in which the elements are passed as arguments)

so your logic for returning true / false also makes sense

amitiuttarwar replied:
some context for why `operator<` is relevant - `std::map` is implemented as a balanced tree (usually red-black tree if you remember that from college cs 😛). so it needs to have the comparison defined to be able to place values on the tree

this allows for the logarithmic insertion / search / deletion operations.

satsie replied:
Had a spare chunk of time that was too small for coding, but just right to catch up on this comment :) Thanks for the resources and the explanation that `std::map` is a self balancing tree. It makes a lot of sense and explains why overloading the `<` operator is necessary!

satsie commented on src/net.cpp:
```diff
@@ -2642,16 +2651,69 @@ void CConnman::RecordBytesRecv(uint64_t bytes)
+    }
+}
+
+void CConnman::PrintNetMsgStats() const {
```
Write an aggregator so I don't have to keep spot checking this with what getnettotals returns

satsie committed:
Remove the = overload

amitiuttarwar commented on src/net.cpp:
```diff
@@ -686,12 +688,18 @@ bool CNode::ReceiveMsgBytes(Span<const uint8_t> msg_bytes, bool& complete, mapMs
+
+            // I'm guessing the C++ here could use some improvement..
+            std::tuple<int, uint64_t> count_bytes = j->second;
+            std::get<0>(count_bytes) += 1;
+            std::get<1>(count_bytes) += msg.m_raw_message_size;
```
UPDATE: this thread has many half-baked ideas, I'm happy to tinker with the C++ at some point to understand better & offer a singular recommendation =P 

----

since the tuple is only used for these two values (count & bytes), `std::pair` seems like it'd offer a cleaner interface. call site would look like this:
```suggestion
            ++count_bytes.first;
            count_bytes.second += msg.m_raw_message_size;
```

amitiuttarwar replied:
oooo, nifty c++ism that could be convenient here -`std::tie` can be used to "unpack" a `std::pair` 

[docs here](https://en.cppreference.com/w/cpp/utility/tuple/tie) - might also be useful to simplify the `operator<` code.

I have to tinker to better understand exactly what the interface would be, but my current understanding is that it could offer something like `std::tie(count, bytes) = j->second;` and then can just use `count` and `bytes` as the variables to increment

UPDATE: Simpler might be better. if we make it a named custom struct, we can name the vars instead of using first & second, or invisible stuff.

satsie replied:
I love it! I actually didn't know the difference between a pair and a tuple so as a first step, I went ahead and converted it `std::pair`. 

It's possible this chunk of code may go away or be refactored when I remove the original stats I added to the `getnettotals` RPC, so I'm going to revisit the `std::tie` suggestion in a little bit :) Interesting that you can name the vars with that data type!

satsie replied:
Can't promise I'll remember the differences between `std::tuple`, `std::pair`, and `std::tie` if this comes up again, but I'll try :) Resolving this comment now that the `count_bytes` data structure no longer exists.

amitiuttarwar commented on src/net.cpp:
```diff
@@ -686,12 +688,18 @@ bool CNode::ReceiveMsgBytes(Span<const uint8_t> msg_bytes, bool& complete, mapMs
             }
-            assert(j != msgtype_bytes.end());
-            j->second += msg.m_raw_message_size;
+            assert(j != msgtype_countbytes.end());
```
for now, this is great. towards the end when we are cleaning up style, I think we can consolidate this logic with the previous `mapRecvBytesPerMsgType` block. they are both initialized to have all valid p2p messages, so they should correlate.

satsie replied:
Was able to completely remove `msgtype_countbytes` so this comment is no longer relevant. Closing!

amitiuttarwar commented on src/net.cpp:
```diff
@@ -1308,19 +1316,20 @@ void CConnman::SocketHandlerConnected(const std::vector<CNode*>& nodes,
-                mapMsgTypeSize msgtype_bytes;
+                // a map to store each message type mapped to the total message count,
+                // as well as the total number of message bytes
+                std::map<std::string, std::tuple<int, uint64_t>> msgtype_countbytes;
```
style note, for later: introduce an alias for this type

much like: `using mapMsgTypeSize = std::map</* message type */ std::string, /* total bytes */ uint64_t>;`

satsie replied:
Closing since `msgtype_countbytes` is no longer needed

amitiuttarwar commented on src/net.cpp:
```diff
@@ -1308,19 +1316,20 @@ void CConnman::SocketHandlerConnected(const std::vector<CNode*>& nodes,
                 }
                 RecordBytesRecv(nBytes);
-                RecordBytesRecvByMsgType(msgtype_bytes);
+                RecordMsgStatsRecv(msgtype_countbytes, pnode->m_conn_type, pnode->ConnectedThroughNetwork());
```
whatcha think about consolidating `RecordBytesRecv` and `RecordMsgStatsRecv`? 

`RecordBytesRecv` actually seems useless in the current state. you could turn it into something useful :)

satsie replied:
Good catch! Addressed in https://github.com/satsie/bitcoin/pull/3/commits/4c3b8216e90596036beda4e1f03d65afd96a3499

To take it a step further, I took a peek at the counterpart method, `RecordBytesSent()` and saw it:
- uses the `m_total_bytes_sent_mutex`, and 
- checks the current time to see if it needs to reset something called a max outbound cycle (`nMaxOutboundCycleStartTime` and `nMaxOutboundTotalBytesSentInCycle`)

After seeing that, I figured it was best to leave `RecordBytesSent()` alone. However, I can see an argument to make `RecordBytesSent()` private, and call it from within `RecordMsgStatsSent()`. Any thoughts on this? Would you consider such a change out of scope?

satsie replied:
**Update**: In commit https://github.com/satsie/bitcoin/pull/3/commits/9322d8a9a9830628cb619ba4c2131b2a22f49fb8 I unfortunately had to undo the change to move `RecordBytesRecv()` into `RecordMsgStatsRecv()`. I was working on [this comment](https://github.com/satsie/bitcoin/pull/3#discussion_r1069234557) to remove the output parameter from `ReceiveMsgBytes()` and ended up having to push the call to record the number of bytes, `RecordMsgStatsRecv()`. 

I am now waiting for the message to be "published" to `CNode`'s `vRecvMsg` list before invoking `RecordMsgStatsRecv()`. Note that `vRecvMsg` only contains completed CNetMessages.

Upon doing this I noticed `test_buffer()` in `p2p_invalid_messsages.py` failing. This is because that test sends a message in two pieces, and `RecordMsgStatsRecv()` doesn't get called until the full message has made it through. Therefore, the `getnettotals`.`totalbytesrecv` value that the test is looking at is not updated until the second half of the message is sent.

Seems like the current behavior is `getnettotals`.`totalbytesrecv` is accurate even when messages are split into pieces, but my new `getnetmsgstats` needs to wait for messages to fully come through before updating. This makes sense to me because I'm collecting message counts, which are whole numbers

amitiuttarwar commented on src/net.h:
```diff
@@ -999,6 +1004,38 @@ class CConnman
     mapMsgTypeSize m_msgtype_bytes_recv;
     mapMsgTypeSize m_msgtype_bytes_sent;
```
just want to clarify - you're planning to remove this previous logic for handling bytes?

satsie replied:
yup! I'm going to circle back at the end. I'm currently using this old code to compare the results from my new code and make sure the byte counts for all the message types are accurate :)

satsie replied:
Old code removed in https://github.com/satsie/bitcoin/pull/3/commits/b459b0f1663c9a71bbdd23ca8808f9527d8c9c9a

Also changed the base branch to master for a cleaner diff

amitiuttarwar commented on src/net.h:
```diff
@@ -550,13 +551,14 @@ class CNode
      */
     bool ReceiveMsgBytes(Span<const uint8_t> msg_bytes, bool& complete,
-                         mapMsgTypeSize& msgtype_bytes) EXCLUSIVE_LOCKS_REQUIRED(!cs_vRecv);
+                         std::map<std::string, std::tuple<int, uint64_t>>& msgtype_bytes) EXCLUSIVE_LOCKS_REQUIRED(!cs_vRecv);
```
variable naming suggestion: 
`msgtype_bytes` -> `msg_stats`

satsie replied:
Since this variable was changed to a `pair`, I prefer the name to reflect the two items in the pair (`msgtype` and `bytes`). But, per one of your earlier comments, if this gets changed to a `tie`, I think calling it a more generic `msg_stats` is the way to go, especially if a `tie` lets you access the elements by names that you define.

satsie replied:
No longer relevant since I reverted the change to add this new parameter to `ReceiveMsgBytes()`. Closing.

amitiuttarwar commented on src/net.h:
```diff
@@ -999,6 +1004,38 @@ class CConnman
     mapMsgTypeSize m_msgtype_bytes_recv;
     mapMsgTypeSize m_msgtype_bytes_sent;
 
+    struct NetMsgStatsKey {
```
variable naming suggestion: `NetMsgStatsKey` -> `MsgStatsKey`
in general I think adding `net` to these variable names isn't really adding any value. and omg there are so many variables that are a mouthful, so I'd love simplifying them :)

satsie replied:
yeah!! Happy to get rid of the "Net" part of the name. Fixed in https://github.com/satsie/bitcoin/pull/3/commits/734ac6fd6c3c3f5f9f395ef629e0fd8e2bc85edc

amitiuttarwar commented on src/net.h:
```diff
@@ -999,6 +1004,38 @@ class CConnman
+
+    struct NetMsgStatsValue {
+        int msg_count;
+        uint64_t num_bytes;
```
variable name suggestions, in theme with simplifying 
```suggestion
    struct MsgStatsValue {
        int count;
        uint64_t bytes;
```

satsie replied:
Similar to the previous comment about `MsgStatsKey` (formerly known as `MsgStatsValue`), I have updated the  name of this struct and dropped the "Net"  https://github.com/satsie/bitcoin/commit/734ac6fd6c3c3f5f9f395ef629e0fd8e2bc85edc :smile: 

I left `msg_count` and actually changed `bytes` to `byte_count`. 
- I think by itself, `count` could possibly be interpreted as the total number of stats being returned. 
- Without looking at the data type, `bytes` could mean the actual bytes themselves. 

Now I can't tell you how realistic it is for someone to misinterpret `count`, or how reasonable it is to make accommodations for readers that don't bother to look at data types. For the latter, I think I have a lot of conditioning from years of working in a dynamically typed language. I'm not sure what the standards are in C++ land, and am happy to drop the "count" off `byte_count` if it's considered acceptable. By that same line of thought, "count"/"msg_count" could then maybe be changed to just `messages` as the int data type would indicate we're not talking about the actual messages themselves.

Though.... as I look at all this I realize when I switch the data structure from a map to an array, none these names should matter. Earlier today I finally fixed a bug in my RPC code that was driving me bananas for days. I was looking for some small wins before diving back into the bigger stuff. I probably should have thought ahead before tackling this naming feedback :grimacing:

amitiuttarwar commented on src/netaddress.cpp:
```diff
@@ -588,6 +588,30 @@ static std::string IPv6ToString(Span<const uint8_t> a, uint32_t scope_id)
     return r;
 }
 
+std::string NetworkAsString(Network network)
```
this works great for now. later you can consider defining a `ToString()` function directly on the `Network`

satsie replied:
Thanks for catching me today before I went down this path :) This comment is resolved by https://github.com/satsie/bitcoin/pull/3#discussion_r1072855453

amitiuttarwar commented on src/net.cpp:
```diff
@@ -1308,19 +1316,20 @@ void CConnman::SocketHandlerConnected(const std::vector<CNode*>& nodes,
+                std::map<std::string, std::tuple<int, uint64_t>> msgtype_countbytes;
 
                 // initialize all message type values to zero
                 for (const std::string& msg : getAllNetMessageTypes())
```
something I don't understand about the existing code is where the infinite loop occurs that keeps the node processing & sending messages for it's whole lifetime. once I understand that, I'll have a better context to evaluate initializing this here. 

(don't worry about this, mostly a note to myself. might be something we ask for feedback from others)

amitiuttarwar replied:
ah, I figured this out (also captured in the send & receive diagram) - `ThreadSocketHandler` calls `SocketHandler` on loop, which goes through each node. so in the end, we wouldn't want to initialize this data structure here every time because of the resource usage. 

one idea is to populate the stats when pulling the data off `vRecvMsg` later in this function, instead of when we are populating the member in `ReceiveMsgBytes`. I believe this should allow us to remove the temporary variable `msgtype_countbytes`.

satsie replied:
Since this comment, I've made updates to hold off on populating stats until later in the function, when the data is pulled off vRecvMsg. 

The reason for `msgtype_countbytes` is that `vRecvMsg` can have multiple messages of the same type. I agree that `msgtype_countbytes` is a larger data structure that I don't want to keep initializing there's another way. The only alternative I see is to call `RecordMsgStatsRecv()` each time a message is processed off of `vRecvMsg` in the loop. 

My gut feeling is this takes less resources but want to get your thoughts on it first.

amitiuttarwar replied:
hm, I don't remember enough context to evaluate off hand, but will keep this in mind to evaluate when I'm reviewing PR#4 :)

satsie replied:
I'm going to follow up on this in response to Vasil's comment below :D https://github.com/satsie/bitcoin/pull/3/files#r1104437828 Closing this one!

amitiuttarwar reviewed (commented):
alrighhht ! I spent a buncha time reviewing the surrounding code around message stats & your code. I now have a solid understanding of both your patch & the existing logic for aggregating stats. 

TLDR; great job ! 

your code is looking really good! you are definitely being thrown off the deep end, but I'm so glad to see you are swimming! in the existing logic, `CConnman` and `CNode` are so intertwined, it took me a while to wrap my head around the aggregated stats vs node stats

I left a LOT of review comments. It's probably a bit overwhelming, but most of them are not things you need to address right now. there's a lot of style notes that you can review once you've finished pouring in the foundation. 

overall the logic looks really strong. you caught the proper touch points of the code, initialize data structures, have things in the right place, etc. once you get the RPC logic wired up, we can tackle locking strategies. add some tests & you will be ready to PR. I think the tests should be pretty easy, since they will be more of sanity checks vs doing anything wild.

these are my outstanding questions about the code design. I don't want you to worry about them right now, I'm just sharing:
1. location of initializing `msgtype_countbytes` (review comment with more detail)
2. what locks should guard the send & receive stats structs
3. follow up -> is it alright for the logic in `ReceiveMsgBytes` to occur while the `cs_vRecv` lock is acquired. that might actually be required, but I want to think it through a bit more & figure out optimizations

oh! I do think it'd be worth removing the old code for v1 of this feature, and having this branch be just your new project against bitcoin master. I think I've got a handle on it all now, but it tripped me up for a while :) 

FYI, I have a bunch of notes about the stats logic on master [here](https://github.com/amitiuttarwar/bitcoin-notes/blob/main/reviews/getnetmsgstats.md)

amitiuttarwar reviewed (commented):
just noting down takeaways from our discussion today: 
- try changing the data structure from two maps to one multi-dimensional array with atomic ints for values. since the size is fixed at compile time, the code doesn't need to lock the whole data structure when reading or writing, and the atomic can provide thread safety for each specific value
- I'll spend some time with the RPC code and let you know what I find

satsie committed:
Aggregate and print the stats

satsie committed:
getnetmsgstats RPC

vasild commented on src/netaddress.cpp:
```diff
@@ -588,6 +588,30 @@ static std::string IPv6ToString(Span<const uint8_t> a, uint32_t scope_id)
     return r;
 }
 
+std::string NetworkAsString(Network network)
```
There is already `std::string GetNetworkName(enum Network net)`.

satsie replied:
Nice! Removed `NetworkAsString()` in https://github.com/satsie/bitcoin/pull/3/commits/0f9a32ec5f561c2d95dd364c850a3f61720a669e

vasild commented on src/net.cpp:
```diff
@@ -1308,19 +1316,20 @@ void CConnman::SocketHandlerConnected(const std::vector<CNode*>& nodes,
+                    msgtype_countbytes[msg] = std::make_tuple(0, 0);
+                msgtype_countbytes[NET_MESSAGE_TYPE_OTHER] = std::make_tuple(0, 0);
 
-                if (!pnode->ReceiveMsgBytes({pchBuf, (size_t)nBytes}, notify, msgtype_bytes)) {
```
There is no need to modify `ReceiveMsgBytes()` to return the extra stats. It has already "published" the relevant data to `vRecvMsg` (if it sets the bool arg (notify/complete) to true). Just below we already iterate over `vRecvMsg` in order to increment `nSizeAdded`. That loop could be extended to gather the stats.

satsie replied:
Oh wow, this is great insight. One thing I wanted to gather from code review was if the output parameter I added to `ReceiveMsgBytes()` was the best solution. Thanks to your comment, I was able to remove the output parameter and instead gather the message type and size in the loop you mentioned just below that iterates over "published" `vRecvMsg` values: https://github.com/satsie/bitcoin/pull/3/commits/9322d8a9a9830628cb619ba4c2131b2a22f49fb8

Combine that with a comment Amiti left about consolidating the functions that record the number of bytes received (see [my update here](https://github.com/satsie/bitcoin/pull/3#discussion_r1084269419)) and I realize this fixes a bug in my code. I had been incrementing the message count even for partial messages. Waiting until a complete message is published to `vRecvMsg` prevents that from happening.

I'm not sure if this warrants a comment in the code, but per [this line,](https://github.com/satsie/bitcoin/pull/3/commits/9322d8a9a9830628cb619ba4c2131b2a22f49fb8#diff-00021eed586a482abdb09d6cdada1d90115abe988a91421851960e26658bed02R1322) `// vRecvMsg contains only completed CNetMessage`, I could note somewhere [around this line](https://github.com/satsie/bitcoin/pull/3/commits/9322d8a9a9830628cb619ba4c2131b2a22f49fb8#diff-00021eed586a482abdb09d6cdada1d90115abe988a91421851960e26658bed02R1327) that `// Messages must be fully received before stats are recorded`. Would that be helpful or noisy? I'm starting to reach a point where things are working enough to consider finer details like which comments are useful, and which ones are redundant.

vasild replied:
No strong opinion. I think it is ok also without a comment `// Messages must be fully received before stats are recorded`.

satsie replied:
Sounds good! I ended up slightly expanding on the existing comment https://github.com/satsie/bitcoin/pull/3/commits/1b5545e31ebd7f92e7b90bf2df4214fc63b778f2

`// vRecvMsg contains only completed CNetMessage, meaning stats are only updated for fully received messages`

vasild commented on src/net.cpp:
```diff
@@ -2642,17 +2651,140 @@ void CConnman::RecordBytesRecv(uint64_t bytes)
+    // ideally this would be JSON but that appears to be hard to do in C++
+    std::map<std::string, NetMsgStatsValue> aggregate_stats = {};
+
+    int num_filters = 0;
+    if (sizeof(filters) > 0) {
+        sizeof(filters) / sizeof(filters[0]);
+    }
```
Maybe you are looking for `const size_t num_filters = filters.size();`?

satsie replied:
yes! :sob: I just discovered this bug independently and fixed. Should have checked the comments here first :sweat_smile: 

Resolved in https://github.com/satsie/bitcoin/pull/3/commits/765bdd48f157ecfb7a36de530db79ce9cb0fafe0#diff-00021eed586a482abdb09d6cdada1d90115abe988a91421851960e26658bed02R2699 and https://github.com/satsie/bitcoin/pull/3/commits/2079a0705fada903917fdef6102f46372b771d09

vasild commented on src/net.cpp:
```diff
@@ -2642,17 +2651,140 @@ void CConnman::RecordBytesRecv(uint64_t bytes)
+    }
+}
+
+std::map<std::string, CConnman::NetMsgStatsValue> CConnman::AggregateNetMsgStats(const std::map<NetMsgStatsKey, NetMsgStatsValue>& raw_stats, std::vector<int> filters) const
```
I don't think this function works correctly. The way I read it, it is going to sum all stats, unconditionally, under a single key of `aggregate_stats`. Anyway, I think it is better to do the aggregation in the RPC code and to return unaggregated stats from `connman` - there you can create directly the output JSON.

I would suggest to leave aggregation alone for now and come up with a completely working and polished code. Aggregation can even be done in a separate followup PR, after the first one which returns unaggregated data is merged.

satsie replied:
As of https://github.com/satsie/bitcoin/pull/3/commits/f5a5d46e0f0aa64a374d817a9e524fcdce792568 aggregation has been moved out of CConnman and into the RPC code. I felt like I was really close with it, so I kept on it, though it did take several revisions :sweat_smile: It is now working and returning a proper JSON object! (not all stats compressed into a single layer of an object)

vasild reviewed (commented):
Thanks for working on this!

Like suggested by @amitiuttarwar, I think it would be better to have a fixed size array. Introduce an enum for the message types and have something like:

```cpp
struct NetMsgStatsValue {
    uint64_t count;
    uint64_t bytes;
};

using Stats = std::array<std::array<std::array<NetMsgStatsValue, NET_MAX>, NUM_CONN_TYPES>, NUM_MSG_TYPES>;

Stats recv
Stats sent;

...

recv[VERACK][ConnectionType::OUTBOUND_FULL_RELAY][NET_IPV4].count = 123;
```

satsie commented:
Thank you so much for taking a peek at this @vasild! :heart: I very much agree with you and Amiti on the fixed size array. We've been chatting about this and it sounds like there may even be some locking benefits too.

I really like your suggestion to get things working without aggregation, and possibly add that as a follow up. I tried to recreate AJ's Python code from [this comment](https://github.com/bitcoin/bitcoin/issues/26337#issuecomment-1343326472) and had a heck of a time trying to implement a map of unknown depth in C++. What I have currently in this PR is a flat map that I just implemented to unblock myself. This aggregation ended up being one of the more challenging parts for me, and I welcome the opportunity to break it off into a separate work unit.

At the moment I'm trying to fix my buggy RPC code. Once that is done I will make a few tests, then circle back to remove the old code from my first iteration of the ticket, and finally start addressing all the wonderful comments and C++ suggestions you and Amiti left. I look forward to getting something in a state that I can share with more people for review. But in the meantime, I am so appreciative of your time and attention  to helping with this, especially as someone who is familiar with this area of the code and the problem that is trying to be solved.

satsie committed:
Use std::pair instead of std::tuple when passing the message count and number of bytes

satsie committed:
Simplify some data structure names

satsie committed:
CConnman: Combine RecordBytesRecv with RecordMsgStatsRecv

satsie committed:
Correct the data type for number of filters

satsie committed:
Remove redundant NetworkAsString()

satsie committed:
Do not record message stats until the message has been published to vRecvMsg

Also restore the separate RecordBytesRecv() setter.
This fixes a bug where message counts were getting incremented for partial messages.

satsie committed:
BUG: fix received stats in RPC response

satsie committed:
Remove updates that added message breakdown to getnettotals

satsie committed:
Improve functional tests

satsie committed:
Replace map with a multidimensional array

ajtowns commented:
> I tried to recreate AJ's Python code from [this comment](https://github.com/bitcoin/bitcoin/issues/26337#issuecomment-1343326472) and had a heck of a time trying to implement a map of unknown depth in C++.

![image](https://user-images.githubusercontent.com/127186/217437777-621609b9-29d8-4832-aa99-d8c701b5f8be.png)

I think in C++ you'd do it something like this (keeping similar-ish naming to what I used in python):

```c++
static constexpr int MSG = 0;
static const std::array<std::string, 2> msgbytes_name = {"msgs", "bytes"};

static const std::string& name = get_name(const std::string& subkey, int msgbytes, int network, int msg, int direction)
{
    if (subkey == "conntype") return conntype_name[direction];
    if (subkey == "network") return network_name[network];
    if (subkey == "msgtype") return msgtype_name[msg];
    if (subkey == "msgbytes") return msgbytes_name[msgbytes];
    throw;
}

UniValue aggregate(
    const std::array<std::array<std::array<std::array<int64_t, 2>, NUM_MSGS>, NUM_NETWORKS>, 2>& x,
    const std::vector<std::string>& keys)
{
    // are we breaking out msgbytes, or just doing byte counts?
    const bool ignore_msgs = std::algorithm::find(keys, [](const auto& x) { return x == "msgbytes"; }); 

    UniValue result{UniValue::VOBJ};
    for (int msgbytes = 0; msgbytes <= 1; ++msgbytes) {
        if (ignore_msgs &&  msgbytes == MSG) continue; // quick skip!

        for (int network = 0; network < NUM_NETWORKS; ++network) {
            for (int msg = 0; msg < NUM_MSGS; ++msg) {
                for (int direction = 0; direction <= 1; ++direction) {
                    const auto v = x[msgbytes][network][msg][direction];

                    // here we're iterating over each individual value (msgbytes, network, msg, direction) as key, v as value

                    UniValue* p = &result;
                    // "result" is our multi-level map, we use p to keep track of where we are as we move down it

                    int levels = keys.size();
                    for (const auto& subkey : keys) {
                        const std::string& name = get_name(subkey, msgbytes, network, msg, direction);
                        // this tells us which subtree we move into

                        if (--levels > 0) {
                            // more sub-levels, still dealing with objects
                            if (!p.exists(name)) p->pushKV(name, UniValue{UniValue::VOBJ});
                            p = &(*p)[name];
                        } else {
                            // final level, we just have an integer
                            if (!p.exists(name)) {
                                p->pushKV(name, v);
                            } else {
                                p = &(*p)[name];
                                p->setInt(p->getInt() + v);
                            }
                        } // if (--levels > 0)
                     } // for (subkey)
                 } // for (direction)
             } // for (msg)
         } // for (network)
    } // for (msgbytes)

    return result;
```            

(not compiled let alone tested; feel free to ping if you want more hints)

satsie commented on src/net.h:
```diff
@@ -853,6 +856,26 @@ class CConnman
 
+    struct MsgStatsValue {
+        int msg_count;
+        uint64_t byte_count;
```
May want to use std::atomic_uint64_t instead. 

Need to make a list of pros and cons (`atomic_uint64_t` vs `uint64_t`)

satsie committed:
Work around so message type enum is not needed

Also fix some variable names from previous commit

satsie committed:
Move initialization of the stats arrays

vasild commented on src/net.cpp:
```diff
@@ -1310,6 +1324,19 @@ void CConnman::SocketHandlerConnected(const std::vector<CNode*>& nodes,
+                        ++count_bytes.first;
+                        count_bytes.second += it->m_raw_message_size;
+
+                        i->second = count_bytes;
```
This and the variable `msgtype_countbytes` can be avoided if stats are directly incremented in `m_netmsg_stats_recv` here. Something like:

```cpp
const size_t net_index = pnode->ConnectedThroughNetwork();
const size_t conn_index = pnode->m_conn_type;
const size_t msg_index = GetMessageTypeIndex(it->m_type);

m_netmsg_stats_recv[net_index][conn_index][msg_index].byte_count += it->m_raw_message_size;
```

Or, even better, put the stats arrays into a class and only access them through a method that ensures no out-of-bounds access happens and also locks anything if needed, should it be decided a mutex will be used for protection. E.g.

```cpp
class NetStats
{
public:
    void RecvInc(Network net, ConnectionType conn_type, const std::string& msg_type, size_t msg_count, size_t byte_count)
    {
        auto& stat = m_recv.at(net).at(conn_type).at(GetMessageTypeIndex(msg_type));
        stat.msg_count += msg_count;
        stat.byte_count += byte_count;
    }

    void SentInc(...
private:
    array m_recv;
    array m_sent;
};
```

And then in `SocketHandlerConnected()`:

```cpp
m_net_stats.Recv(pnode->ConnectedThroughNetwork(), pnode->m_conn_type, it->m_type, /*msg_count=*/1, it->m_raw_message_size);
```

satsie replied:
You read my mind!!! Thank you! I was recently looking at this in the context of a comment that Amiti left: https://github.com/satsie/bitcoin/pull/3#discussion_r1104718777

Wasn't sure if it was worth getting rid of `msgtype_countbytes` completely, but I had a suspicion that this was the way to go. Also appreciate the `NetStats` class idea :fire: I moved the initialization of the arrays into this class as well. Hope that's the right place. 

The rest of my changes for this are in https://github.com/satsie/bitcoin/pull/3/commits/7952c020e5551a0d7e9a9be5f180a403816bef90

vasild commented on src/net.cpp:
```diff
@@ -1345,7 +1373,16 @@ void CConnman::SocketHandlerConnected(const std::vector<CNode*>& nodes,
             // Send data
             size_t bytes_sent = WITH_LOCK(pnode->cs_vSend, return SocketSendData(*pnode));
-            if (bytes_sent) RecordBytesSent(bytes_sent);
+            if (bytes_sent) {
+                RecordBytesSent(bytes_sent);
+                // RecordMsgStatsSent(msg.m_type, bytes_sent, pnode->m_conn_type, pnode->ConnectedThroughNetwork());
+
+                // TODO how to get the message type here? This code will be hit if CConnman::PushMessage has messages on the queue
+                // that get sent now by SocketSendData.
+                // One option is to just record as sent once CConnman::PushMessage() puts them on the CNode.vSendMsg queue.
+                // Somewhere around this line in CConnman::PushMessage():
+                //        pnode->mapSendBytesPerMsgType[msg.m_type] += nTotalSize;
+            }
```
Yes, I think there is no need to do anything here and we should only record it as `mapSendBytesPerMsgType[]` even though it would mean that we record some bytes as sent before them being actually sent (and they may never be sent). That seems acceptable given the complexity otherwise, if we want to record after the send. And this is how `mapSendBytesPerMsgType[]` works as well.

satsie replied:
Thanks for taking a look at this! I have moved the code that updates the sent message stats over to CConnMan::PushMessage() in this commit: https://github.com/satsie/bitcoin/pull/3/commits/1b5545e31ebd7f92e7b90bf2df4214fc63b778f2

vasild commented on src/net.cpp:
```diff
@@ -2808,7 +2971,10 @@ void CConnman::PushMessage(CNode* pnode, CSerializedNetMsg&& msg)
         if (optimisticSend) nBytesSent = SocketSendData(*pnode);
     }
-    if (nBytesSent) RecordBytesSent(nBytesSent);
+    if (nBytesSent) {
+        RecordBytesSent(nBytesSent);
+        RecordMsgStatsSent(msg.m_type, nBytesSent, pnode->m_conn_type, pnode->ConnectedThroughNetwork());
+    }
```
I guess this should happen unconditionally, even if nothing was sent. Just next to where `pnode->mapSendBytesPerMsgType[msg.m_type]` is updated above (and be the only place where the sent recording is done).

satsie committed:
Simplify code and remove msgtype_countbytes

Move the stats arrays into a NetStats class. To prevent out-of-bounds access, only allow access via methods.
Also fix a bug where the *other* message type was not being handled

satsie committed:
Record sent stats in CConnman::PushMessage

satsie committed:
add comment

satsie committed:
Aggregation - Phase I, get aggregation out of CConnman

satsie commented:
@ajtowns Thank you for the code hints! The `UniValue` was the big thing that I was missing. I had been trying to do it with nested maps :sweat_smile: 

I am (continuing to fall into the trap and) trying to work with the new sample code you provided. I am getting stuck towards the end at this part:
```
                            if (!p.exists(name)) p->pushKV(name, UniValue{UniValue::VOBJ});
                            path = &(*p)[name];
```

According to the [docs](https://doxygen.bitcoincore.org/class_uni_value.html), the `operator[]` returns a `const UniValue &`. I can't seem to find a way to get what is in `p[name]` in a way that is not const. Surely there's some C++ reference/pointer magic I'm not understanding here, right? I believe I will later need to push a key-value to `p[name]` which is why I can't work with something that is a const. 

This is what my current attempt to get past this looks like, but I haven't been able to prove if it works. `path` was not defined in your latest code snippet, but I believe the intent is to update `p`, so the last line of my code here sets `p` to a new value. Will continue hacking on it.
```
                        UniValue new_path{UniValue::VOBJ};
                        if (!p->exists(filter_name)) p->pushKV(filter_name, new_path);
                        p = &new_path;
```

Is this the right track? Can you let me know your thoughts on this? Thank you for your help!

ajtowns commented:
> I am (continuing to fall into the trap and) trying to work with the new sample code you provided. I am getting stuck towards the end at this part:

Hmm, you're right; `getKeys` and `getValues` are also unhelpfully const, and `getObjMap` creates copies of the objects so also isn't useful. 

Probably means you have to do your own `map` after all. Something like:

```c++
#include <util/overloaded.h>

class MultiLevelMap
{
private:
    size_t m_levels = 0;
    std::variant<
        std::map<std::string, int64_t>,
        std::map<std::string, MultiLevelMap>> m_varmap;

public:
    explicit MultiLevelMap(size_t levels) : m_levels{levels}
    {
        if (levels > 1) {
            m_varmap.emplace<1>();
        } else {
            m_varmap.emplace<0>();
        }
    }

    int64_t* get_value(const Span<std::string>& steps)
    {
        if (steps.size() == 0) return nullptr;
        return std::visit(util::Overloaded{
            [&](std::map<std::string, int64_t>& m) {
                return &m[steps.front()]; // create if not present
            },
            [&](std::map<std::string, MultiLevelMap>& m) {
                size_t sublevel = m_levels > 0 ? m_levels - 1 : 0;
                auto [it, inserted] = m.try_emplace(steps.front(), sublevel);
                return it->second.get_value(steps.subspan(1));
            }}, m_varmap);
    }

    UniValue to_univalue() const
    {
        UniValue result;
        std::visit(util::Overloaded{
            [&](const std::map<std::string, int64_t>& m) {
                for (auto& [k,v] : m) {
                    result.pushKV(k, v);
                }
            },
            [&](const std::map<std::string, MultiLevelMap>& m) {
                for (auto& [k,v] : m) {
                    result.pushKV(k, v.to_univalue());
                }
            }}, m_varmap);
        return result;
    }
};
```

would, I think, let you say:

```c++
     MultiLevelMap x{4};

     // loop over all the items, generating the "path" for each one:
     std::vector<std::string> path{"bytes", "ping", "inbound", "ipv4"};
     int64_t* v = x.get_value(path);
     if (v != nullptr) *v += 1024;

     // then convert to univalue for conversion to json
     return x.to_univalue();
```

(Moving the `get_name()` logic from above into `get_value()` would avoid having to update the path vector constantly)

> Is this the right track? Can you let me know your thoughts on this? Thank you for your help!

Would have been the right track if UniValue had some non-const accessors for its values, but I couldn't see any either.

satsie committed:
Aggregation - Phase II, use multi level map

satsie committed:
Fix tests by updating RPC return types

satsie commented:
@ajtowns thank you thank you thank you! It works! In my earlier attempt to do this with maps I recall learning about variants, but didn't feel confident enough in my understanding to do a custom MultiLevelMap implementation like the one you provided. I would not have been able to do that on my own.

I updated the sample request/responses in the PR description up top and it's so beautiful to see the results nested into a pretty JSON object 🥲

satsie committed:
Remove helper code that is no longer needed

satsie commented on src/rpc/net.cpp:
```diff
@@ -506,6 +508,301 @@ static RPCHelpMan getaddednodeinfo()
+                                        "the statistics for that type are 0. Only known message,\n"
+                                        "connection, and network types can appear as keys in the object.",
+                {
+                    // TODO create a helper function similar to MempoolEntryDescription() in mempool.cpp
```
Note to self: try and do this during PR 4

satsie commented on src/test/fuzz/rpc.cpp:
```diff
@@ -131,6 +131,7 @@ const std::vector<std::string> RPC_COMMANDS_SAFE_FOR_FUZZING{
     "gettxspendingprevout",
     "getmempoolinfo",
     "getmininginfo",
+    "getnetmsgstats",  // TODO does this belong here?
```
Do not bring this over to PR 4, but make a comment about it

(Or do some homework between now and then to get the answer)

satsie committed:
Pass message count through, but do not do anything with it yet

satsie commented on src/net.cpp:
```diff
@@ -1087,6 +1085,9 @@ bool CConnman::AddConnection(const std::string& address, ConnectionType conn_typ
         break;
+    // conn_type should never be set to NUM_CONN_TYPES
+    case ConnectionType::NUM_CONN_TYPES:
+        break; // TODO is break okay here? or assert(false)?
```
Leave comment in PR 4

satsie commented on src/protocol.cpp:
```diff
@@ -47,6 +47,9 @@ const char *WTXIDRELAY="wtxidrelay";
 } // namespace NetMsgType
 
+// TODO: moved this from net.cpp. Should it be in the NetMsgType namespace? If so, what about allNetMessageTypes?
+const std::string NET_MESSAGE_TYPE_OTHER = "*other*";
```
leave TODO in PR 4

satsie commented on src/rpc/net.cpp:
```diff
@@ -509,6 +511,301 @@ static RPCHelpMan getaddednodeinfo()
+                    // get the filter_name, which tells us which subtree to move into
+                    if (filter == 0) {
+                        // message type
+                        // Should NET_MESSAGE_TYPE_OTHER be added to getAllNetMessageTypes?
```
leave comment in PR 4

satsie commented on test/functional/rpc_net.py:
```diff
@@ -168,12 +172,97 @@ def test_getnettotals(self):
+        assert_array_result([getnetmsgstats_conntype_msgtype['received']['manual']['pong']], {'msg_count': 2}, {'total_bytes': 64})
+
+        getnetmsgstats_conntype_msgtype_network = self.nodes[0].getnetmsgstats(filters=["conntype", "msgtype", "network"])
+        # TODO: is it okay to test against the "not_publicly_routable" network type?
```
comment in PR 4?

satsie commented on test/functional/rpc_net.py:
```diff
@@ -168,12 +172,97 @@ def test_getnettotals(self):
+        getnetmsgstats_conntype_msgtype_network = self.nodes[0].getnetmsgstats(filters=["conntype", "msgtype", "network"])
+        # TODO: is it okay to test against the "not_publicly_routable" network type?
+        manualconn_sendcmpct_stats = [{'msg_count': 2}, {'total_bytes': 66}]
+        assert_array_result([getnetmsgstats_conntype_msgtype_network['sent']['manual']['sendcmpct']['not_publicly_routable']],
```
This line always fails after a fresh build, but never subsequently

satsie commented:
Got a few open comments here that I need to move elsewhere, will close this PR in favor of #4 soon.
